### PR TITLE
Add a description key to theme.json style variations

### LINKED ISSUE #44960: Add a description field to theme.json style variations
## What problem does this address?

Sometimes a name or title is not enough to describe the changes in a style variation. Especially if the user has a difficulty using the visual miniature preview.

## What is your proposed solution?

Add an additional description field to the theme.json style variations.
The description would be used in the global styles sidebar, and for the preview of the style variations in the wordpress.org theme directory.

### PR DESCRIPTION
**Pull request by carolinan**
## What?

- Adds an optional description key to theme.json, theme-i18n.json, and theme.json schema.
- The description is added to the aria label of the style variation container. If the style variation has no description, only the title is used.

Closes https://github.com/WordPress/gutenberg/issues/44960

## Why?

This is intended to be an accessibility improvement. Style variations are mostly visual, and there needs to be a text alternative that describes them.
One purpose is to use the text description to improve the style variation preview in the WordPress.org theme directory.

## Testing Instructions
Install the theme Twenty Twenty-Three.
In your code editor, locate and open the Aubergine.json style variation in the themes `styles `folder.
Add the new description key and value as a top-level key in the JSON file:
`"description": "Your description here"`
```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"title": "Aubergine",
	"description": "This is the test description"
```
Activate the theme.
Go to Appearance, Editor.
Activate your screen reader if you haven't already.
In the Navigation Sidebar, select Styles from the menu.
Using the tab key, navigate past the Back button, the Stylebook button, and the Edit Styles button.
Navigate past the first style variation button, which is called Default.
The second button should announce the name "Aubergine" followed by the description you entered.

Next, please test the style variations in the edit mode:
Use the Shift + tab to navigate backward until you reach the Edit Styles button. Activate the button.
Tab forward until you reach the Stylebook button; this button means you have reached the Styles sidebar.
Tab forward past the Style Actions button until you reach Browse Styles. Activate this button.
Navigate past the first style variation button, which is called Default.
The second button should announce the name "Aubergine" followed by the description you entered.

## Screenshots or screencast 
https://github.com/WordPress/gutenberg/assets/7422055/b1355998-c1e5-4036-a3b0-7a2c8df72797

### CODE REVIEW

**Comment by carolinan** [comment]
@ramonjd Hi, I have a question about an e2e test.

There is an e2e test for the [style variation screen](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/specs/site-editor/style-variations.spec.js) where a style variation is selected. The test checks for a [button](https://github.com/WordPress/gutenberg/blob/trunk/test/e2e/specs/site-editor/style-variations.spec.js#L77) that contains the style variation **title**.

The test passes right now because I added in the condition that if the style variation file has no description, only the title is used in the aria-label of the button.
The style variations used in the tests does not have a description.

Must the test be updated to also check for a button that contains the description?

**Comment by ramonjd** [comment]
Hi @carolinan 

Thanks for the ping.

> The test passes right now because I added in the condition that if the style variation file has no description, only the title is used in the aria-label of the button.
> The style variations used in the tests does not have a description.

I think it's fine to leave the e2e test as it is. I don't think the test was meant to cover what the content of the label is in this instance, but rather how and whether the style variations interface is working. In that way `await page.click( 'role=button[name="pink"i]' );` is just a means to an end.

> Must the test be updated to also check for a button that contains the description?

I'd say, and this is just my opinion, that testing whether `title + description` is working as expected would be better suited for a unit test if you think it's important to cover it.

File: packages/edit-site/src/components/global-styles/screen-style-variations.js [hunk 1]
```diff
@@ -72,6 +72,10 @@ function Variation( { variation } ) {
 		return compareVariations( user, variation );
 	}, [ user, variation ] );
 
+	const label =
+		variation.title +
+		( variation?.description ? '. ' + variation?.description : '' );
```
Comment (carolinan) [inline packages/edit-site/src/components/global-styles/screen-style-variations.js]: I am hesitant about combining the variation title (name) and description this way.
Should both be inside the aria-label? Perhaps the description needs to be separate?

Secondly, perhaps a non issue, but they are both translatable texts and there is no indication for translators that the texts will be combined this way.

Comment (carolinan) [reply packages/edit-site/src/components/global-styles/screen-style-variations.js]: Discussed this with @afercia and the better way would be to use aria-describedby for the description.

File: packages/edit-site/src/components/global-styles/screen-style-variations.js [hunk 2]
```diff
@@ -96,6 +103,11 @@ function Variation( { variation } ) {
 						isFocused={ isFocused }
 						withHoverView
 					/>
+					{ variation?.description && (
```
Comment (youknowriad) [inline packages/edit-site/src/components/global-styles/screen-style-variations.js]: Should this use a VisuallyHidden component?

**Comment by youknowriad** [comment]
I wonder if the description should actually be visible. Maybe some design thoughts would be good on this PR.

**Comment by carolinan** [comment]
@jasmussen Hi! Who can help decide if the style variations in the styles sidebar should have a visible description?

**Comment by carolinan** [comment]
@youknowriad After the WP 6.0 compatibility files were removed, I seem to be unable to re-add "description" as a new top level key in theme.json in either the WP_Theme_JSON_6_1 or WP_Theme_JSON_6_2 classes (I tried both).
This means that the description field is no longer found inside "variation" and the description is not being output.
Help wanted :)

**Comment by youknowriad** [comment]
I think you need to copy the functions that make use of the updated constants in the 6.2 file or something in that vein.

**Comment by jasmussen** [comment]
Descriptions for style variations is not something that's yet been designed for the style variations, and so the UI hasn't been built to scale to them being exposed. Given that the style variation UI is likely going to be very much in flux in the 6.2 phase (see #45371) I'd be concerned about exposing such a description for the small thumbnails. 

To be clear, I think there's could be a place for it. For example, right now this big thumbnail looks clickable, even if in fact it does nothing:

Screenshot 2022-12-13 at 10 18 38

I could see a design where clicking that surfaces a description as a flyout, perhaps metadata or other items. But it would need a bit of a design iteration. In that vein, I wouldn't visually surface these until we have a better plan for them.

**Comment by carolinan** [comment]
I will wait until #46579 is resolved then.

**Comment by carolinan** [comment]
I am not sure what I am missing. The updated `VALID_TOP_LEVEL_KEYS` constant in plugins/gutenberg/lib/class-wp-theme-json-gutenberg.php is not being used.
If I add `description` to `VALID_TOP_LEVEL_KEYS` in core wp-includes/class-wp-theme-json.php, the PR works.

File: packages/edit-site/src/components/global-styles/screen-style-variations.js (hunk 1 above)
Comment (carolinan) [reply packages/edit-site/src/components/global-styles/screen-style-variations.js]: The aria-label overrides the aria-describedby, so the description is not announced.

**Comment by carolinan** [comment]
> I am not sure what I am missing. The updated `VALID_TOP_LEVEL_KEYS` constant in plugins/gutenberg/lib/class-wp-theme-json-gutenberg.php is not being used. If I add `description` to `VALID_TOP_LEVEL_KEYS` in core wp-includes/class-wp-theme-json.php, the PR works.

This has been resolved. The PR description and testing instructions have also been updated.

**Review by ajlende — APPROVED** [review]
I don't see any reason we can't merge this and have a follow-up to add it to the UI once the design has been made.

The schema looks good, and I see the description in the label. 👍

**Comment by oandregal** [comment]
@carolinan @ajlende @ramonjd what's the status of the new string `Description`? Is it available for users in the UI / needs translation?

While reviewing the `theme-i18n.json` files, I've noticed the string is present in the [gutenberg file](https://github.com/WordPress/gutenberg/blob/91d7e4173a266e1ee9a5f5d23321cd2d58b7e8c0/lib/theme-i18n.json#L3) but it's not 1) in the [core one](https://github.com/WordPress/wordpress-develop/blob/5effffccd005ca3dae6631ee73b1096b8f54a861/src/wp-includes/theme-i18n.json#L2) (this one defines the strings to be extracted for translation) and 2) the [i18n-command file](https://github.com/wp-cli/i18n-command/blob/6f52398030618d6e3436d0ba533fbeb5458ebf04/assets/theme-i18n.json#L2) (acts as a fallback and needs to be in sync with core). It seems that [the core backport PR](https://github.com/WordPress/wordpress-develop/pull/4687) missed updating that file.

If the string is displayed to users in any way, we'd need to prepare PRs to update core and i18n-command filse ([see example for i18n-command](https://github.com/wp-cli/i18n-command/pull/405)). If it doesn't require translation, we should remove it from the gutenberg's file.

This has landed as part of WordPress 6.3, so I don't think there's any urgency to address it for 6.6 (we're in RC), though it'd be good to clarify in the coming days/weeks.

**Comment by carolinan** [comment]
I did not know about the command file.

The description is part of the aria-label for the style variations In Editor > Styles and in 
Site Editor > Styles Sidebar > Browse Styles.

It is not visible but used to add context for screen reader users. It needs to be translatable.

**Comment by ramonjd** [comment]
> It is not visible but used to add context for screen reader users. It needs to be translatable.

In that case, would the process be:

1. Add the `"description"` field to [theme-i18n.json](https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/theme-i18n.json#L2), maybe under `"title"`?
2. Get the above patch committed
3. Prep the i18n-command file patch using https://github.com/wp-cli/i18n-command/pull/405 as an example?


_Edit_: Prepped a Core PR here: [Trac ticket](https://core.trac.wordpress.org/ticket/61543) | [Core PR ](https://github.com/WordPress/wordpress-develop/pull/6947)

I found an example from @oandregal that added the top-level `"title"` key https://github.com/wp-cli/i18n-command/pull/306

I have a draft PR up here: https://github.com/wp-cli/i18n-command/pull/408

I'm not sure if it's right.

**Comment by oandregal** [comment]
Thanks for the follow-ups. I approved the core PR and the i18n-command PR is also working well for me (I cannot approve/merge there so I pinged the relevant people).

**Pull request merged**